src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -470,7 +493,11 @@ private TypeSymbol NarrowObjectType(TypeValidatorConfig config, ObjectSyntax exp
                         originSyntax: config.OriginSyntax,
                         onTypeMismatch: GetPropertyMismatchDiagnosticWriter(config, ShouldWarn(targetType), declaredProperty.Name));
 
-                    var narrowedType = NarrowType(newConfig, declaredPropertySyntax.Value, declaredProperty.TypeReference.Type);
+                    // append "| null" to the property type for non-required properties
+                    var (propertyAssignmentType, typeWasPreserved) = AddImplicitNull(declaredProperty.TypeReference.Type, declaredProperty.Flags);
+
+                    var narrowedType = NarrowType(newConfig, declaredPropertySyntax.Value, propertyAssignmentType);
+                    narrowedType = RemoveImplicitNull(narrowedType, typeWasPreserved);
```

Comment:
>RemoveImplicitNull [](http://example.com/codeflow?start=35&length=18)

Removal is needed because we are using the assigned type for things that should ideally be using the declared type.